server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/OrderedShardsIteratorTests.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.routing.allocation.allocator;
+
+import org.elasticsearch.cluster.ESAllocationTestCase;
+import org.elasticsearch.cluster.node.DiscoveryNodes;
+import org.elasticsearch.cluster.routing.IndexRoutingTable;
+import org.elasticsearch.cluster.routing.RoutingNodes;
+import org.elasticsearch.cluster.routing.RoutingTable;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.index.Index;
+import org.elasticsearch.index.shard.ShardId;
+import org.hamcrest.Description;
+import org.hamcrest.TypeSafeMatcher;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+
+import static org.elasticsearch.cluster.routing.ShardRoutingState.STARTED;
+import static org.elasticsearch.cluster.routing.TestShardRouting.newShardRouting;
+import static org.hamcrest.Matchers.allOf;
+import static org.hamcrest.Matchers.anyOf;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.not;
+
+public class OrderedShardsIteratorTests extends ESAllocationTestCase {
+
+    public void testOrdersShardsAccordingToAllocationRecency() {
+
+        var nodes = DiscoveryNodes.builder().add(newNode("node-1")).add(newNode("node-2")).add(newNode("node-3")).build();
+
+        var routing = RoutingTable.builder()
+            .add(index("index-1a", "node-1"))
+            .add(index("index-1b", "node-1"))
+            .add(index("index-2", "node-2"))
+            .add(index("index-3", "node-3"))
+            .build();
+
+        var ordering = new NodeAllocationOrdering();
+        ordering.recordAllocation("node-1");
+
+        var iterator = OrderedShardsIterator.create(RoutingNodes.mutable(routing, nodes), ordering);
```

Comment:
nit: let us use immutable instead, seems more inline with the read-only nature of the iterator:

```suggestion
        var iterator = OrderedShardsIterator.create(RoutingNodes.immutable(routing, nodes), ordering);
```

Reply:
While iterator is not performing any changes I wanted to test it it with mutable RoutingNodes as this is what we supply it in the real code